### PR TITLE
Fixes #7936 Check failure when writing an uploaded file

### LINKED ISSUE #7936: File upload fails without exception if disk is full
When I upload a file to Play server, if the disk has not enough space to store the file, the file is silently truncated.
No exception is raised so application can't show appropriate error message to user.

My environment:
 - Play Scala 2.6.3
 - Oracle Java 1.8.0_144
 - Linux

### PR DESCRIPTION
**Pull request by To-om**
# Pull Request Checklist

* [X] Have you read [How to write the perfect pull request](https://github.com/blog/1943-how-to-write-the-perfect-pull-request)?
* [X] Have you read through the [contributor guidelines](https://www.playframework.com/contributing)?
* [X] Have you signed the [Lightbend CLA](https://www.lightbend.com/contribute/cla)?
* [X] Have you referenced any issues you're fixing using [commit message keywords](https://help.github.com/articles/closing-issues-using-keywords/)?
* [X] Have you added copyright headers to new files?
* [ ] Have you checked that both Scala and Java APIs are updated?
* [ ] Have you updated the documentation for both Scala and Java sections?
* [X] Have you added tests for any changed functionality?

# Helpful things

Fixes #7936 

## Purpose

The PR check IOResult provided by FileIO.toPath when a temporary file is created.
If file creation fails, the error is reported to error handler.

### CODE REVIEW

File: framework/src/play/src/test/scala/play/api/mvc/InMemoryTemporaryFileCreator.scala [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+import com.github.sbridges.ephemeralfs.EphemeralFsFileSystemBuilder
+import play.api.libs.Files.{ TemporaryFile, TemporaryFileCreator }
+
+import scala.util.Try
+
+class InMemoryTemporaryFile(val path: Path, val temporaryFileCreator: TemporaryFileCreator) extends TemporaryFile {
+  def file: File = path.toFile
+}
+
+class InMemoryTemporaryFileCreator(totalSpace: Long) extends TemporaryFileCreator {
```
Comment (gmethvin) [inline framework/src/play/src/test/scala/play/api/mvc/InMemoryTemporaryFileCreator.scala]: nice 👍

File: framework/src/play/src/main/scala/play/core/parsers/Multipart.scala [hunk 2]
```diff
@@ -113,7 +114,9 @@ object Multipart {
 
       }
 
-      multipartAccumulator.through(handleFileParts)
+      multipartAccumulator.through(handleFileParts).recoverWith {
+        case t => errorHandler.onServerError(request, t).map(Left(_))
```
Comment (gmethvin) [inline framework/src/play/src/main/scala/play/core/parsers/Multipart.scala]: I actually think a failed accumulator is a better way of representing this error case. A  `Left` here indicates an error in parsing the actual contents of the body. What happens if we remove this `recoverWith`?

Comment (To-om) [reply framework/src/play/src/main/scala/play/core/parsers/Multipart.scala]: I don't know if an exception thrown in body parser is catched and sent to error handler. If yes this `recoverWith` is not useful.
If it is removed, my test need to be fixed with something like `Await.result(response, Duration.Inf) must throwA[IOException]`

File: framework/build.sbt [hunk 3]
```diff
@@ -50,7 +50,8 @@ lazy val PlayJodaFormsProject = PlayCrossBuiltProject("Play-Joda-Forms", "play-j
 lazy val PlayProject = PlayCrossBuiltProject("Play", "play")
     .enablePlugins(SbtTwirl)
     .settings(
-      libraryDependencies ++= runtime(scalaVersion.value) ++ scalacheckDependencies ++ cookieEncodingDependencies,
+      libraryDependencies ++= runtime(scalaVersion.value) ++ scalacheckDependencies ++ cookieEncodingDependencies :+
+        "com.github.sbridges" % "ephemeralfs" % "1.0.1.0" % Test,
```
Comment (marcospereira) [inline framework/build.sbt]: Declare `"com.github.sbridges" % "ephemeralfs" % "1.0.1.0" % Test` in framework/project/Dependencies.scala and then use it here.

File: framework/build.sbt [hunk 4]
```diff
@@ -50,7 +50,8 @@ lazy val PlayJodaFormsProject = PlayCrossBuiltProject("Play-Joda-Forms", "play-j
 lazy val PlayProject = PlayCrossBuiltProject("Play", "play")
     .enablePlugins(SbtTwirl)
     .settings(
-      libraryDependencies ++= runtime(scalaVersion.value) ++ scalacheckDependencies ++ cookieEncodingDependencies,
+      libraryDependencies ++= runtime(scalaVersion.value) ++ scalacheckDependencies ++ cookieEncodingDependencies :+
+        ephemeralfs % Test,
 
       sourceGenerators in Compile += Def.task(PlayVersion(
         version.value,
```
Comment (cchantep) [inline framework/build.sbt]: Is it really required?

File: framework/project/Dependencies.scala [hunk 5]
```diff
@@ -161,6 +161,8 @@ object Dependencies {
 
   val cookieEncodingDependencies = slf4j
 
+  val ephemeralfs = "com.github.sbridges" % "ephemeralfs" % "1.0.1.0"
```
Comment (schmitch) [inline framework/project/Dependencies.scala]: shouldn't we use jimfs here? https://github.com/google/jimfs ephermeralfs looks really outdated.

Comment (cchantep) [reply framework/project/Dependencies.scala]: That's also a concern to me. Or maybe some utils from the already imported guava

File: framework/src/play/src/test/scala/play/api/mvc/MultipartBodyParserSpec.scala [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+import play.api.http.{ HttpEntity, HttpErrorHandler }
+import play.core.test.{ FakeHeaders, FakeRequest }
+
+import scala.concurrent.{ Await, Future }
+import scala.concurrent.duration.Duration
+import scala.util.Random
+
+class MultipartBodyParserSpec extends Specification {
+
+  "Multipar body parser" should {
```
Comment (marcospereira) [inline framework/src/play/src/test/scala/play/api/mvc/MultipartBodyParserSpec.scala]: `Multipar` -> `Multipart` (missing `t` at the end).

**Comment by marcospereira** [comment]
@To-om thank you so much for this.

I've added an integration test so we can better test the fix.

@gmethvin, @schmitch and @cchantep you may want to review this again.

**Comment by marcospereira** [comment]
Thank you, @To-om. See #8043 for a backport to 2.6.x.

**Pull request merged**